src/v2/Components/LocationAutocompleteInput.tsx

```diff
@@ -0,0 +1,161 @@
+  onChange(place: Place): void
+}
+
+export const LocationAutocompleteInput: React.FC<LocationAutocompleteInputProps> = ({
+  onChange,
+  ...rest
+}) => {
+  const [ready, setReady] = useState(false)
+
+  useLoadScript({
+    id: "google-maps-js",
+    src: GOOGLE_PLACES_API_SRC,
+    onReady: () => {
+      setReady(true)
+    },
+  })
+
+  useAppendStylesheet({
+    id: "google-maps-css",
+    body: `.pac-container { z-index: ${DEFAULT_MODAL_Z_INDEX}; }`,
+  })
+
+  const inputRef = useRef<HTMLInputElement | null>(null)
+  const autocompleteServiceRef = useRef<any | null>(null)
+
+  useEffect(() => {
+    if (!ready || !inputRef.current) return
+
+    // @ts-ignore
+    autocompleteServiceRef.current = new google.maps.places.Autocomplete(
+      inputRef.current,
+      { types: ["(cities)"] }
+    )
+
+    autocompleteServiceRef.current.addListener("place_changed", () => {
+      const place: Place | undefined = autocompleteServiceRef.current.getPlace()
+      if (!place) return
+      onChange(place)
+    })
+
+    // @ts-ignore
+    google.maps.event.addDomListener(
+      inputRef.current,
+      "keydown",
+      (event: KeyboardEvent) => {
+        if (event.key === "Enter") {
+          event.preventDefault()
+        }
+      }
+    )
+  }, [onChange, ready])
+
+  return <Input ref={inputRef as any} {...rest} />
+}
+
+type AddressComponent = {
+  long_name: string
+  short_name: string
+  types: string[]
+}
+
+export type Place =
+  /** When Google can't match a city, it simply returns the name entered */
+  | { name: string }
+  /** Typically Google returns this rich place object */
+  | {
+      address_components: AddressComponent[]
+      adr_address: string
+      formatted_address: string
+      geometry: {
+        location: {
+          lat(): number
+          lng(): number
+        }
+        viewport: {
+          south: number
+          west: number
+          north: number
+          east: number
+        }
+      }
+      icon: string
+      icon_background_color: string
+      icon_mask_base_uri: string
+      name: string
+      photos: {
+        height: number
+        html_attributions: string[]
+        width: number
+      }[]
+      place_id: string
+      reference: string
+      types: string[]
+      url: string
+      utc_offset: number
+      vicinity: string
+      html_attributions: any[]
+      utc_offset_minutes: number
+    }
+
+export type Location = {
+  city: string
+  state?: string
+  stateCode?: string
+  postalCode?: string
+  country?: string
+  coordinates?: number[]
+}
+
+export const normalizePlace = (place: Place): Location => {
+  if (!("address_components" in place)) {
+    return { city: place.name }
+  }
+
+  const components = place.address_components.reduce(
+    (
+      acc: {
+        state?: AddressComponent
+        postalCode?: AddressComponent
+        country?: AddressComponent
+      },
+      component
```

Comment:
totally non-blocking - i know there's the `address_components` array from google, but maybe we can use a less overloaded term here. Devs might see `component.foo` below and think we're doing some funny business with react

Reply:
Yeah, I had the same though. 👍